### PR TITLE
SE: Upload processing

### PR DESCRIPTION
**Pull request by ahmad-elbatanouni**
1. Send a notification to the coordinator
2. Update the Task

### CODE REVIEW

**Comment by ahmad-elbatanouni** [comment]
[SE: Upload processing](https://trello.com/c/r0CHM9Yi/28-se-upload-processing)

File: functions/src/modules/SE.functions.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+//      2. Update the Task
+//
+/////////////////////////////////////////////////
+
+export const uploadProcessing = functions.storage.object()
+.onFinalize(async object => {    
+
+    const filePath = object.name;
+    
+    if (object.name.slice(-4).toLowerCase() !== 'flac') {
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply functions/src/modules/SE.functions.ts]: Used the mimetype `contentType` instead

File: functions/src/modules/SE.functions.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    const filePathRegex = /sound-editing\/restored\/(\w+)\/(\w+)/g;
+
+    if (!filePath.match(filePathRegex)) {
+        console.error("Something is wrong with the file path!")
+        return -1;
+    }
+    
+    const match = filePathRegex.exec(filePath);
+    const list = match[1], fileNameFlac = match[2];
+    const taskId = fileNameFlac.slice(0, -5); // strip the trailing ".flac" extenstion
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply functions/src/modules/SE.functions.ts]: there's an attribute on the file object called `contentType` ... do you think it's better to modify the regex to match only files ending in **.flac**?

File: functions/src/modules/SE.functions.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    const filePathRegex = /sound-editing\/restored\/(\w+)\/(\w+)/g;
+
+    if (!filePath.match(filePathRegex)) {
+        console.error("Something is wrong with the file path!")
+        return -1;
+    }
+    
+    const match = filePathRegex.exec(filePath);
+    const list = match[1], fileNameFlac = match[2];
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply functions/src/modules/SE.functions.ts]: I tried using this (package)[https://www.npmjs.com/package/named-regexp-groups] but it seems to have some issues.
Do you have a particular suggestion .. or should I move on testing more packages ?

File: functions/src/modules/SE.functions.ts [hunk 4]
```diff
@@ -16,15 +16,13 @@ export const uploadProcessing = functions.storage.object()
...
-
     const filePathRegex = /sound-editing\/restored\/(\w+)\/(\w+)/g;
 
-    if (!filePath.match(filePathRegex)) {
-        console.error("Something is wrong with the file path!")
+    if (!filePath.match(filePathRegex)) 
+        return -1;    
+
+    if (object.contentType === 'audio/flac') {
+        console.error("Something is wrong with the file extension!")
```
Comment (anantakrishna) [inline functions/src/modules/SE.functions.ts]: The message should mention the file type, not an extension. The file can have FLAC content type and another extension.

File: functions/src/modules/SE.functions.ts [hunk 5]
```diff
@@ -50,11 +47,11 @@ export const uploadProcessing = functions.storage.object()
     // 2. Update the Task
     let taskRestorationUpdate = {
         status: 'Uploaded',
-        timestampLastVersion: Math.round((new Date()).getTime() / 1000)
+        timestampLastVersion: admin.database.ServerValue.TIMESTAMP
     };
 
     if (!task.restoration.timestampFirstVersion)
-        taskRestorationUpdate['timestampFirstVersion'] = Math.round((new Date()).getTime() / 1000);
+        taskRestorationUpdate['timestampFirstVersion'] = admin.database.ServerValue.TIMESTAMP;
```
Comment (anantakrishna) [inline functions/src/modules/SE.functions.ts]: Can't we use a property assignment syntax here?

File: functions/src/modules/SE.functions.ts [hunk 6]
```diff
@@ -34,14 +32,13 @@ export const uploadProcessing = functions.storage.object()
```
Comment (anantakrishna) [inline functions/src/modules/SE.functions.ts]: Use named capture groups to get the list and the task id here.

File: functions/src/modules/SE.functions.ts [hunk 7]
```diff
@@ -16,15 +16,13 @@ export const uploadProcessing = functions.storage.object()
...
-        console.error("Something is wrong with the file extension!")
-        return -1;
-    }
-
     const filePathRegex = /sound-editing\/restored\/(\w+)\/(\w+)/g;
 
-    if (!filePath.match(filePathRegex)) {
-        console.error("Something is wrong with the file path!")
+    if (!filePath.match(filePathRegex)) 
+        return -1;
```
Comment (anantakrishna) [inline functions/src/modules/SE.functions.ts]: Just check for the path prefix here to filter out the events. Regex should be more specific and should used after this simple check.

File: functions/src/modules/SE.functions.ts [hunk 8]
```diff
@@ -16,15 +16,13 @@ export const uploadProcessing = functions.storage.object()
 
     const filePath = object.name;
     
-    if (object.name.slice(-4).toLowerCase() !== 'flac') {
-        console.error("Something is wrong with the file extension!")
-        return -1;
-    }
-
     const filePathRegex = /sound-editing\/restored\/(\w+)\/(\w+)/g;
```
Comment (anantakrishna) [inline functions/src/modules/SE.functions.ts]: `\w` is not good here. Test for task ID and the file extension in this regex.

File: functions/src/modules/SE.functions.ts (hunk 6 above)
Comment (ahmad-elbatanouni) [reply functions/src/modules/SE.functions.ts]: It's not supported yet, we need to use a package .. I tried using this [package](https://www.npmjs.com/package/named-regexp-groups) but it seems to have some issues.
Do you have a particular suggestion .. or should I move on testing more packages ?

File: functions/src/modules/SE.functions.ts (hunk 5 above)
Comment (ahmad-elbatanouni) [reply functions/src/modules/SE.functions.ts]: It's not a property of the **taskRestorationUpdate** object, so the compiler would complain if used  as a property.
Also I'm adding it only the first time the `restoration` sub-node is updated .. so I can't add a  placeholder property of the same name to the **taskRestorationUpdate** object.

File: functions/src/modules/SE.functions.ts (hunk 8 above)
Comment (ahmad-elbatanouni) [reply functions/src/modules/SE.functions.ts]: Ok, but the grouping still needs to be based on the whole name `list-serial.flac` .. right ?

File: functions/src/modules/SE.functions.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (anantakrishna) [reply functions/src/modules/SE.functions.ts]: Oh, I see, ECMAScript does not support named capture groups. Never mind then.

**Pull request closed without merging**